doc/mgmt/SONiC_YANG_Model_Guidelines.md

```diff
@@ -0,0 +1,961 @@
+| XPath          | XML Path Language              |
+| CVL		 | Configuration Validation Library |
+
+
+## Overview
+ This document lists the guidelines, which will be used to write YANG Modules for SONiC. These YANG Modules (called SONiC YANG models) will be primarily based on or represent the ABNF.json of SONiC, and the syntax of YANG models must follow RFC 7950 ([https://tools.ietf.org/html/rfc7950](https://tools.ietf.org/html/rfc7950)).  Details of config in format of ABNF.json can be found at https://github.com/Azure/SONiC/wiki/Configuration.
+
+These YANG models will be used to verify the configuration for SONiC switches, so a library which supports validation of configuration on SONiC Switch must use these YANG Models. List of such Libraries are: 1.) Configuration Validation Library. (CVL).  YANG models, which are written using these guidelines can also be used as User End YANG Models, i.e North Bound configuration tools or CLI can provide config data in sync with these YANG models. 
+
+[SONiC Management Framework](https://github.com/Azure/SONiC/pull/436) uses SONiC YANG models for configuration management in two ways :
+
+* As *custom Northbound Management YANG* in absence of standard YANG model and for configuration validation (CVL uses it).
+* As *CVL YANG* for configuration validation only, when standard YANG model (Openconfig, IETF etc.) is used as Northbound Management YANG. In such case, steps below which are marked as "This step applies when SONiC YANG is used as Northbound YANG." can be skipped.
+
+
+## Guidelines
+
+
+### 1. All schema definitions related to a feature should be written in a single YANG model file. YANG model file is named as 'sonic-{feature}.yang', e.g. for ACL feature sonic-acl.yang is the file name. It is mandatory to define a top level YANG container named as 'sonic-{feature}' i.e. same as YANG model name. For example, sonic-acl.yang should have 'sonic-acl' as top level container. All other definition should be written inside the top level container.
+
+Example :
+####  YANG
+```
+module sonic-acl {
+	container sonic-acl {
+		.....
+		.....
+	}
+}
+```
+
+
+### 2.  It is best to categorize YANG Modules based on a networking components. For example, it is good to have separate modules for VLAN,  ACL, PORT and IP-ADDRESSES etc.
+```
+sonic-acl.yang
+sonic-interface.yang
+sonic-port.yang
+sonic-vlan.yang
+```
+
+### 3. Define namespace as "http://github.com/Azure/{model-name}" and define a shorter(preferably within 3-4 letters) namespace prefix. Use the prefix in XPath for referring to any node present in other model.
+
+Example :
+####  YANG
+```
+module sonic-acl {
+	namespace "http://github.com/Azure/sonic-acl";
+	prefix sacl;
+	.....
+	.....
+}
+```
+
+### 4. Use 'revision' to record revision history whenever YANG model is changed. Updating revision with appropriate details helps tracking the changes in the model.
+
+Example :
+
+#### YANG
+```
+module sonic-acl {
+	revision 2019-09-02 {
```

Comment:
Let`s keep the length of Example minimum possible. If it is about revision, please show only revision. Applicable to all examples in this PR.

Reply:
Ok.